### PR TITLE
Add scheduling checks to ajax scheduler

### PR DESCRIPTION
**Pull request by willgearty**
This adds a set of scheduling checks to the ajax scheduler that can be enabled by selecting them in the new options tab (only one can be enabled at a time):
image

Once a scheduling check is enabled, it will be used to color the cells in the scheduling matrix. For example, for the "Double-Booked Teachers" check, it will color sections red that overlap in time and have the same teacher(s):
image

This functionality works for looking at sections that are already scheduled and for scheduling sections in new spots (e.g. if the capacity check is enabled, the color of the cell will change as you move your cursor over different cells in the matrix based on the capacities of the classrooms). This means that in addition to using these checks AFTER scheduling, you can also use them DURING scheduling to see if adding/moving a section in a particular spot will cause or remove an issue (e.g. will make a teacher "hungry" or have to "run"). You can see this in this example screenshot where I am moving the B6s1 section:
image

Then, once a section is scheduled, we update the colors of ALL cells to indicate if new issues have arisen. Note: I'm not sure how the performance is for this last part for larger programs, but I imagine it shouldn't be too bad since javascript is pretty efficient, and it's not like it will prevent you from scheduling other sections.

Fixes https://github.com/learning-unlimited/ESP-Website/issues/1406.

### CODE REVIEW

File: esp/public/media/scripts/ajaxschedulingmodule/ESP/Cell.js [hunk 1]
```diff
@@ -96,13 +100,175 @@ function Cell(el, section, room_id, timeslot_id, matrix) {
...
+                                        continue teacher_loop;
+                                    }
+                                }
+                            }
+                        }
+                }
+                if(n_teachers == 0) return grey;
+                return this.cellColors.HSLToRGB(0,100,100 - 10 * n_teachers); // Color red based on number of teachers
+            case "double_booked":
+                // Count how many teachers are teaching other sections at the same time as this one
```
Comment (kkbrum) [inline esp/public/media/scripts/ajaxschedulingmodule/ESP/Cell.js:214]: is this referring to coteachers of a class? I think that could be more clear if so.

File: esp/public/media/scripts/ajaxschedulingmodule/ESP/Cell.js [hunk 2]
```diff
@@ -96,13 +100,175 @@ function Cell(el, section, room_id, timeslot_id, matrix) {
...
+                                    continue teacher_loop;
+                                }
+                            }
+                        }
+                        if(n_teachers == 5) break teacher_loop;
+                    }
+                if(n_teachers == 0) return grey;
+                return this.cellColors.HSLToRGB(0,100,100 - 10 * n_teachers); // Color red based on number of teachers
+            case "running":
+                // Count how many teachers are teaching other sections immediately before or after this one in a different room
```
Comment (kkbrum) [inline esp/public/media/scripts/ajaxschedulingmodule/ESP/Cell.js:232]: ditto

**Review by kkbrum — CHANGES_REQUESTED** [review]
Really cool feature addition.

Not sure what's going on, but for both the "running teachers" and "double-booked teachers", one of my classes turns black and is no longer clickable: 

image

**Comment by willgearty** [comment]
Can you take a screenshot of the console log when that happens @kkbrum?

**Review by kkbrum — APPROVED** [review]
A little hard to test with my limited classes, but it looks good and I'm excited to see it in action!

**Pull request merged**